types/vote_set.go

```diff
@@ -671,6 +672,110 @@ func (voteSet *VoteSet) MakeExtendedCommit(fp FeatureParams) *ExtendedCommit {
 	return ec
 }
 
+// MakeBLSCommit is a variant of MakeExtendedCommit that aggregates BLS signatures.
+//
+// It additionally aggregates the BLS signatures for the block and nil. The
+// resulting commit contains only two aggregated signatures:
+//
+// 1 - aggregated signature for the block (compressed)
+// 2 - aggregated signature for nil (compressed).
+//
+// Note the signatures count is preserved, but only the first signature in the
+// each category (block, nil) is non-empty.
+func (voteSet *VoteSet) MakeBLSCommit() *ExtendedCommit {
+	voteSet.mtx.Lock()
+	defer voteSet.mtx.Unlock()
+
+	if voteSet.signedMsgType != PrecommitType {
+		panic("Cannot MakeBLSCommit() unless VoteSet.Type is PrecommitType")
+	}
+
+	// Make sure we have a 2/3 majority
+	if voteSet.maj23 == nil {
+		panic("Cannot MakeBLSCommit() unless a blockhash has +2/3")
+	}
+
+	// 1. Aggregate the signatures for the block.
+	sigsToAgg := make([][]byte, 0, len(voteSet.votes))
+	for _, v := range voteSet.votes {
+		if v != nil && v.BlockID.IsComplete() {
+			// if block ID exists but doesn't match, exclude sig
+			if !v.BlockID.Equals(*voteSet.maj23) {
+				continue
+			}
+			sigsToAgg = append(sigsToAgg, v.Signature)
+		}
+	}
+	agSig1, err := bls12381.AggregateSignatures(sigsToAgg)
+	if err != nil {
+		panic(fmt.Errorf("BLS aggregation error: %w", err))
+	}
+
+	// 2. Aggregate the signatures for nil, if any.
+	sigsToAgg = make([][]byte, 0, len(voteSet.votes))
+	for _, v := range voteSet.votes {
+		if v != nil && v.BlockID.IsNil() {
+			sigsToAgg = append(sigsToAgg, v.Signature)
+		}
+	}
+	var agSig2 []byte
+	if len(sigsToAgg) > 0 {
+		agSig2, err = bls12381.AggregateSignatures(sigsToAgg)
+		if err != nil {
+			panic(fmt.Errorf("BLS aggregation error: %w", err))
+		}
+	}
+
+	// For every validator, get the precommit without extensions
+	sigs := make([]ExtendedCommitSig, len(voteSet.votes))
+	for i, v := range voteSet.votes {
+		cSig := v.CommitSig()
+		if cSig.BlockIDFlag != BlockIDFlagAbsent {
+			cSig.Signature = []byte{0x00} // clear the signature
+		} else {
+			cSig.Signature = []byte{} // clear the signature
+		}
+		sig := ExtendedCommitSig{
+			CommitSig: cSig,
+		}
+		// if block ID exists but doesn't match, exclude sig
+		if sig.BlockIDFlag == BlockIDFlagCommit && !v.BlockID.Equals(*voteSet.maj23) {
```

Comment:
But you just declare `sig` in line 738, and didn't set any flag at this point...